mrmustard/lab/representations/data/matvec_data.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import numpy as np
+
+from mrmustard.lab.representations.data.data import Data
+from mrmustard.math import Math
+from mrmustard.physics.gaussian import reorder_matrix_from_qpqp_to_qqpp
+from mrmustard.typing import Matrix, Scalar, Vector
+
+math = Math()
+
+
+class MatVecData(Data):  # Note: this class is abstract!
+    r"""Contains matrix and vector -like data for certain Representation objects.
+
+    Args:
+        mat: the matrix-like data to be contained in the class
+        vec: the vector-like data to be contained in the class
+        coeffs: the coefficients
+    """
+
+    def __init__(self, mat: Matrix, vec: Vector, coeffs: Scalar) -> None:
+        self.mat = mat  # math.atleast_3d(mat)
+        self.vec = vec  # math.atleast_2d(vec)
+        self.coeffs = coeffs  # math.atleast_1d(coeffs)
+
+    def __neg__(self) -> MatVecData:
+        new_coeffs = -self.coeffs
+        return self.__class__(self.mat, self.vec, new_coeffs)
+
+    def __eq__(self, other: MatVecData) -> bool:
+        try:
+            return super().same(
+                X=[self.mat, self.vec, self.coeffs], Y=[other.mat, other.vec, other.coeffs]
+            )
+
+        except AttributeError as e:
+            raise TypeError(f"Cannot compare {self.__class__} and {other.__class__}.") from e
+
+    def __add__(self, other: MatVecData) -> MatVecData:
+        try:
+            if super().same(X=[self.mat, self.vec], Y=[other.mat, other.vec]):
+                combined_coeffs = self.coeffs + other.coeffs
+                return self.__class__(self.mat, self.vec, combined_coeffs)
+
+            else:
+                mat = math.concat([self.mat, other.mat], axis=0)
+                vec = math.concat([self.vec, other.vec], axis=0)
+                reorder_matrix = reorder_matrix_from_qpqp_to_qqpp(self.mat.shape[-1])
+                mat = math.matmul(math.matmul(reorder_matrix, mat), math.transpose(reorder_matrix))
```

Comment:
I have no idea of this matmul...

Reply:
OK, so what should we do about this? Shall we ask Seb/Filippo to review it too?

Reply:
Sorted